src/controller/artifact/processor/default.go

```diff
@@ -16,44 +16,178 @@ package processor
+	"encoding/json"
+	"io/ioutil"
+	"net/url"
 	"regexp"
 	"strings"
+
+	schemaV1 "github.com/goharbor/harbor/src/controller/artifact/schema/v1"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/pkg/artifact"
+	"github.com/goharbor/harbor/src/pkg/registry"
+
+	"github.com/docker/distribution/manifest/schema2"
+	"github.com/mitchellh/mapstructure"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
 // ArtifactTypeUnknown defines the type for the unknown artifacts
 const ArtifactTypeUnknown = "UNKNOWN"
 
 var (
+	// DefaultProcessor is to process artifact which has no specific processor
+	DefaultProcessor = &defaultProcessor{RegCli: registry.Cli}
+
 	artifactTypeRegExp = regexp.MustCompile(`^application/vnd\.[^.]*\.(.*)\.config\.[^.]*\+json$`)
 )
 
 // the default processor to process artifact
 // currently, it only tries to parse the artifact type from media type
 type defaultProcessor struct {
-	mediaType string
+	RegCli registry.Client
 }
 
-func (d *defaultProcessor) GetArtifactType() string {
+func (d *defaultProcessor) GetArtifactType(ctx context.Context, artifact *artifact.Artifact) string {
 	// try to parse the type from the media type
-	strs := artifactTypeRegExp.FindStringSubmatch(d.mediaType)
+	strs := artifactTypeRegExp.FindStringSubmatch(artifact.MediaType)
 	if len(strs) == 2 {
 		return strings.ToUpper(strs[1])
 	}
 	// can not get the artifact type from the media type, return unknown
 	return ArtifactTypeUnknown
 }
-func (d *defaultProcessor) ListAdditionTypes() []string {
-	return nil
+func (d *defaultProcessor) ListAdditionTypes(ctx context.Context, artifact *artifact.Artifact) []string {
+	artifactConfig := artifact.ExtraAttrs
+	schemaMap, ok := artifactConfig[schemaV1.XHarborAttributes]
+	if !ok {
+		return nil
+	}
+	schema := &schemaV1.Schema{}
+	err := mapstructure.Decode(schemaMap, schema)
+	if err != nil {
+		log.Warningf("unsupported artifact config schema: %v", err)
+		return nil
+	}
+	additionTypes := make([]string, 0)
+	for _, addition := range schema.Additions {
+		additionTypes = append(additionTypes, addition.Name)
+	}
+	return additionTypes
 }
-func (d *defaultProcessor) AbstractMetadata(ctx context.Context, manifest []byte, artifact *artifact.Artifact) error {
-	// do nothing currently
-	// we can extend this function to abstract the metadata in the future if needed
+func (d *defaultProcessor) AbstractMetadata(ctx context.Context, artifact *artifact.Artifact, manifest []byte) error {
+	if artifact.ManifestMediaType != v1.MediaTypeImageManifest && artifact.ManifestMediaType != schema2.MediaTypeManifest {
+		return nil
+	}
+	// get manifest
+	mani := &v1.Manifest{}
+	if err := json.Unmarshal(manifest, mani); err != nil {
+		log.Warningf("unsupported manifest: %v", err)
+		return nil
+	}
+	// get config layer
+	_, blob, err := d.RegCli.PullBlob(artifact.RepositoryName, mani.Config.Digest.String())
+	if err != nil {
```

Comment:
Not very understand return nil error even we get a real error and the return of the func is error type?

Maybe return an error and log all the error (as a warning) at the place of calling this interface method?

Reply:
If we return real error, it will block artifact pushing to registry. I think whatever user defiend in artifact, we shouldn't block it pushing to registry.

Reply:
@ywk253100 Is that correct?

What I mean here is we can handle the error at the place where we call this func. If we think it should not block the pushing process, we can just log error and by pass it at the calling place.

```go
// ensure the artifact exists under the repository, create it if doesn't exist.
func (c *controller) ensureArtifact(ctx context.Context, repository, digest string) (bool, *artifact.Artifact, error) {
	art, err := c.artMgr.GetByDigest(ctx, repository, digest)
	// the artifact already exists under the repository, return directly
	if err == nil {
		return false, art, nil
	}

	// got other error when get the artifact, return the error
	if !errors.IsErr(err, errors.NotFoundCode) {
		return false, nil, err
	}

	// the artifact doesn't exist under the repository, create it first
	repo, err := c.repoMgr.GetByName(ctx, repository)
	if err != nil {
		return false, nil, err
	}
	artifact := &artifact.Artifact{
		ProjectID:      repo.ProjectID,
		RepositoryID:   repo.RepositoryID,
		RepositoryName: repository,
		Digest:         digest,
		PushTime:       time.Now(),
	}
	// abstract the metadata for the artifact
	if err = c.abstractor.AbstractMetadata(ctx, artifact); err != nil {
	// HERE WE LOG?
              log.warnning(error)
              // return false, nil, err
	}
     
        // ... ignore the left part
}
```

Reply:
I prefer to return the error and block the pushing operation.
If users define the customized configurations, what they expect is that the configuration items can be managed and shown in Harbor.
And fail early is helpful for users to realize the error and build their artifacts in Harbor standard way

Reply:
> I prefer to return the error and block the pushing operation.
> If users define the customized configurations, what they expect is that the configuration items can be managed and shown in Harbor.
> And fail early is helpful for users to realize the error and build their artifacts in Harbor standard way

OK, you got me. I will return real error. Thanks.